astropy/modeling/tests/test_quantities_evaluation.py

```diff
@@ -0,0 +1,538 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""
+Tests that relate to evaluating models with quantity parameters
+"""
+
+from __future__ import (absolute_import, unicode_literals, division,
+                        print_function)
+
+from numpy.testing import assert_allclose
+
+
+from ..core import Fittable1DModel
+from ..parameters import Parameter
+from ..models import Gaussian1D
+from ... import units as u
+from ...units import UnitsError
+from ...tests.helper import pytest, assert_quantity_allclose
+
+
+# We start off by taking some simple cases where the units are defined by
+# whatever the model is initialized with, and we check that the model evaluation
+# returns quantities.
+
+
+@pytest.mark.xfail
+def test_evaluate_with_quantities():
+    """
+    Test evaluation of a single model with Quantity parameters that do
+    not explicitly require units.
+    """
+
+    # We create two models here - one with quantities, and one without. The one
+    # without is used to create the reference values for comparison.
+
+    g = Gaussian1D(1, 1, 0.1)
+    gq = Gaussian1D(1 * u.J, 1 * u.m, 0.1 * u.m)
+
+    # We first check that calling the Gaussian with quantities returns the
+    # expected result
+    assert_quantity_allclose(gq(1 * u.m), g(1) * u.J)
+
+    # Units have to be specified for the Gaussian with quantities - if not, an
+    # error is raised
+    with pytest.raises(UnitsError) as exc:
+        gq(1)
+    assert exc.value.args[0] == ("Units of input 'x', (dimensionless), could not be "
+                                 "converted to required input units of m (length)")
+
+    # However, zero is a special case
+    assert_quantity_allclose(gq(0), g(0) * u.J)
+
+    # We can also evaluate models with equivalent units
+    assert_allclose(gq(0.0005 * u.km).value, g(0.5))
+
+    # But not with incompatible units
+    with pytest.raises(UnitsError) as exc:
+        gq(3 * u.s)
+    assert exc.value.args[0] == ("Units of input 'x', s (time), could not be "
+                                 "converted to required input units of m (length)")
+
+    # We also can't evaluate the model without quantities with a quantity
+    with pytest.raises(UnitsError) as exc:
+        g(3 * u.m)
+    assert exc.value.args[0] == ("Input in m (length), could not be converted "
+                                 "to required to dimensionless input")
```

Comment:
I'm -0 on this. I can also see that a model defined without units just acts like any other function. After all
```
def example(x):
    return x**2
```
will work for `x = 3` and for `x= 3 * u.m`. I've come to view models as functions (although they have more machinery under the hood), so I might expect that they work like functions. Not all models will work (e.g. a model that tries to do `x + 3` will fail is `x` is a quantity) but again, that's just like any other function.